### DISCUSSION TITLE
Negative requirement to reject a wiring? How to force a particular Java version?

### DISCUSSION THREAD

**Discussion by mickaelistria** [eclipse-equinox/equinox#492] [Q&A] [answered]
I'm working on a plugin that use javac internal API. This API is not really an API, so it changes on every Java version.
I'd like to express that my bundle is stuck to a particular version of Java, which I know has the necessary Javac API. At the moment, just restraining to 1 particular major version of Java seems enough (we may need finer grain later, but not yet, so let's not bother with it unless it brings a simpler solution right now).
If I use `Bundle-RequiredExecutionEnvironment: JavaSE-21` then Java 22 will match
If I use `Require-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version=21))"` then Java 22 will match because it does provide the Java 22 capability
If I use ``Require-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version=21)(version<22))" then Java 22 would still match, because its osgi.ee 21 capability will match.

Is there a way to tell Equinox "I don't want to wire against any module which provides this capability? Something like `Reject-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version>21)(version<22))` ?

Or maybe it's something to reconsider more deeply in Equinox: maybe Java 22 should *not* provide osgi,ee 21 capability and should *only* provide the osgi.ee 22 one, and a `BREE: JavaSE-21` which seems currently translated to `Require-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version=21)` shoudl instead be translated to `Require-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version>=21)` ?
I believe this would allow the simpler cases to keep working while allowing finer grain restrictions against particular Java versions.

**Comment by laeubi** [comment]
osgi.ee / bree are more a way to express the used class version not says anything about the java packages version so I think this is correct. If you need specific java features, you should therefore import the java packages you use.

For this particular very special case you can use a Multi-Release jar and just add a dummy implementation that throws an exception for any java version higher than the supported one, that way you can even support multiple java versions in the same bundle:

- https://github.com/eclipse-tycho/tycho/blob/main/RELEASE_NOTES.md#building-multi-release-jars

**Reply by mickaelistria** [reply]
> you should therefore import the java packages you use.

OK, it makes sense.
And can we use version constraints for these packages? IIRC the system packages are not versions, so I cannot tell `Import-Package: com.sun.tools.javac.parser; version="[21.22)"`, can I?

**Reply by laeubi** [reply]
No you can't version these java packages, your ee defines the minimum version and if you use preview API you either need to constrain your deployment or use multi release jars.

**Reply by mickaelistria** [reply]
Multi-release jar can be intersting if we want to support multiple Java versions at once though; thanks for sharing this tip!
In the meantime, I guess I'll just add an activator that crashes the `start` method if Java version is > to the one defined in the requirement.

**Reply by laeubi** [reply]
Just be aware that the bundle will still resolve!

**Reply by laeubi** [reply]
Thinking more about it... you can have a Multirelease jar and craft the Manifest by hand (as shared in the link) the have an unsolvable constraint  for higher java version for example:

```
Provide-Capability: osgi.unresolvable
Require-Capability: osgi.unresolvable;filter:="(&(must.not.resolve=*)(
 !(must.not.resolve=*)))"
```

**Reply by mickaelistria** [reply]
That seems perfect, thank you!

**Reply by merks** [reply]
The JREs packages are not versioned. 

I wonder if  if something like this works?
```
(& (osgi.ee=JavaSE) (version=21)(!(version>21)))
```
So it must have version 21 **and** it must **not** have a version higher than 21.

**Reply by mickaelistria** [reply]
> `(& (osgi.ee=JavaSE) (version=21)(!(version>21)))`

This would still bind to the `(osgi.ee=JavaSE) (version=21)` capability that is provided by eg Java 22 (because Java N provide JavaSE-K capability for any K <= N.

@laeubi The multi-release jar trick is really promising. Before adopting it, I have 1 question: are the MANIFEST entries ovewritten by specific versions, or are they expanded?
Ie n my case, if I have in MANIFEST.MF `Require-Capability: osgi.unresolvable;filter:="(&(must.not.resolve=*)(
 !(must.not.resolve=*)))"` and in _META-INF/versions/21/OSGI-INF/MANIFEST.MF_ `Require-Capability: osgi.ee;filter:=" (& (osgi.ee=JavaSE) (version=21))"`, does the later one override the former one or does it "append" it (so it wouldn't resolve?

**Reply by laeubi** [reply]
https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module-multireleasejar

> the Framework **must replace the values of the following manifest headers** in the manifest with the values of these headers, if present, in the supplemental manifest file

So it is always a replacement if you specify it, by the way you might just choose a better name e.g. `can.only.work.on.java.21`

**Reply by laeubi** [reply]
You can take a look at the example e.g.

https://github.com/eclipse-tycho/tycho/blob/main/demo/multi-release-jar/META-INF/versions/11/OSGI-INF/MANIFEST.MF

will replace the import package an minimum java version here:

https://github.com/eclipse-tycho/tycho/blob/main/demo/multi-release-jar/META-INF/MANIFEST.MF

**Reply by merks** [reply]
@mickaelistria 

I would test your assertion because the way I understand it, the one thing providing these capabilities, i.e., the system bundle, must satisfy all the constraints specified.  So as I understand it you can specify that the system bundle must have a certain constraint version of the capability.  I think you are assuming that the capabilities themselves is isolated things that can be used on their own when in fact they are bound to the thing providing them.  I.e., this is about what are the valid wires to the system bundle...

**Reply by laeubi** [reply]
@merks I don't think the system bundle exports one requirement with a version list but multiple ones with individual versions, so it will at least match one of them it is not required that it matches all (what would be impossible e.g. if I require Java 11 at minimum and the framework provides also Java 8 then this would be the same situation).

**Reply by merks** [reply]
I think this works according to this test:

![image](https://github.com/eclipse-equinox/equinox/assets/208716/caf792e3-f3ab-418c-bf00-3fdf28b5f943)

I.e., the system bundle just have version 16 but may not have version 17 where my actual JRE is version 17 in the launch.  Of course any JRE 17 or higher will have the 17 capability and will be rejected...

**Reply by merks** [reply, marked as answer]
Actually testing the runtime behavior I see this failure:
```
org.osgi.framework.BundleException: Could not resolve module: org.eclipse.oomph.setup.core [403]
  Unresolved requirement: Require-Bundle: org.example.test; bundle-version="1.0.0"
    -> Bundle-SymbolicName: org.example.test; bundle-version="1.0.0.qualifier"
       org.example.test [551]
         Unresolved requirement: Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=16)(!(version=17)))"

	at org.eclipse.osgi.container.Module.start(Module.java:493)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:2069)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:143)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:2060)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:2000)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1963)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1879)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:341)
```
So I do think this should work and does work.

**Reply by merks** [reply]
FYI, the system bundles provides these capabilities:

![image](https://github.com/eclipse-equinox/equinox/assets/208716/dbefaa3f-28be-40d0-974c-7921e1f07909)

**Reply by mickaelistria** [reply]
> So I do think this should work and does work.

OK, great thanks. I was certain it wouldn't work for the reason mentioned by @laeubi (I had some confused memories of the requirements being separated). But it's really great that it does.
Thank you very much for verifying my assertions and proving me wrong; I'm in practice happier that way ;)

**Reply by merks** [reply]
Helpful people like you have earned helpfulness in return.   🥇   And I really did want to understand this better myself too!
